### DISCUSSION TITLE
Parsing a null-terminated UTF-16 string?

### DISCUSSION THREAD

**Discussion by landaire** [winnow-rs/winnow#497] [Q&A]
I've been considering rewriting my binary format parser from hand-rolled reads using `byteorder` to something `winnow`-based, and thought I'd try my hand at something that seems pretty straightforward: reading a null-terminated UTF-16 string. Please note, I have no prior experience with winnow but do have some with `nom`.

Currently what I'm trying to do is:

```rust
fn utf16_null_terminated(mut input: Partial<&[u8]>) -> PResult<String> {
    let (str_bytes, _term): (Vec<u16>, _) =
        repeat_till(0.., be_u16, seq!(0, 0)).parse_next(&mut input)?;

    Ok(String::from_utf16(&str_bytes).expect("failed to convert UTF-16 string"))
}
```

This doesn't *feel* very idiomatic for some reason? I think that to me, I expected to be able to match on the data type I'm parsing *to* -- so instead of `seq!(0, 0)`, I initially tried `repeat_till(0.., be_u16, 0u16)`, which obviously didn't work.

So two questions:

1. Is there a more preferred way to achieve this?
2. For the sake of argument, let's say that I have a more complex data type with its own terminator marker -- like a flag. How would I capture this scenario (or the one I provided above where I was trying to simply match the parsed u16).

**Comment by epage** [comment]
>  expected to be able to match on the data type I'm parsing to -- so instead of seq!(0, 0), I initially tried repeat_till(0.., be_u16, 0u16), which obviously didn't work.

The challenge with this is knowing its endianness.  Kind of like [Caseless](https://docs.rs/winnow/latest/winnow/ascii/struct.Caseless.html), I could see us having `Endian::Big(T)` type that would work like `""`, being a parser directly, working in `literal`, and [`take_until`](https://docs.rs/winnow/latest/winnow/token/fn.take_until.html).

> Is there a more preferred way to achieve this?

This seems reasonable.
- You could use `.map()` for converting to the string.
- I'd like use `(0, 0)` or `b"\0\0"` over `seq!(0, 0)`
- Are all all u16's valid in UTF-16?  I'd assume its like UTF-8 and there is a chance for error.
- Your parser doesn't take `input` as `&mut` which is the standard signature needed to implement the `Parser` trait and you could lose track of where you are at within a document

> For the sake of argument, let's say that I have a more complex data type with its own terminator marker -- like a flag. How would I capture this scenario (or the one I provided above where I was trying to simply match the parsed u16).

I'm not too sure what you are looking for with this.  For what I understand of the question, your code does this.
